### PR TITLE
feat: mls remove user from mls group FS-546

### PR DESCRIPTION
**Pull request by ghuMustafa**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/FS-546" title="FS-546" target="_blank">TaskFS-546</a>  [iOS] Admin removes a user from an MLS group
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

- Added `removeMembersFromConversation` method in `MLSMLSController`.
- Added the call of in `ZMConversation+Participants` and handled the case for proteus and mls.

### Testing

- Added tests for `removeMembersFromConversation` in `MLSMLSController`.
----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: Source/Model/Conversation/ZMConversation+Participants.swift [hunk 1]
```diff
@@ -200,9 +201,46 @@ extension ZMConversation {
...
+        case .proteus:
+            var action = RemoveParticipantAction(user: user, conversation: self)
+            action.onResult(resultHandler: completion)
+            action.send(in: context.notificationContext)
+
+        case .mls:
+            guard
+                let mlsController = context.mlsController,
+                let groupID = mlsGroupID?.base64EncodedString,
+                let mlsGroupID = MLSGroupID(base64Encoded: groupID)
```
Comment (johnxnguyen) [inline Source/Model/Conversation/ZMConversation+Participants.swift:215]: `mlsGroupID` should already be of type `MLSGroupID`. So you only need to unwrap it from the conversation:

`guard let groupID = mlsGroupID else { ... )`

File: MLS/MLSController.swift [hunk 2]
```diff
@@ -33,6 +33,7 @@ public protocol MLSControllerProtocol {
 
     func addMembersToConversation(with users: [MLSUser], for groupID: MLSGroupID) async throws
 
+    func removeMembersFromConversation(with clientIds: [ClientId], for groupID: MLSGroupID) async throws
```
Comment (johnxnguyen) [inline MLS/MLSController.swift:36]: Instead of passing in `ClientId` which is just an alias for bytes, we could pass in the more descriptive `[MLSClientID]`, then extract the bytes. This is in line with the responsibility of the `MLSController` to be a high level interface to the low level `CoreCrypto`.

File: Source/Model/Conversation/ZMConversation+Participants.swift [hunk 3]
```diff
@@ -200,9 +201,46 @@ extension ZMConversation {
...
+            Task {
+                do {
+                    try await mlsController.removeMembersFromConversation(with: mlsClientIDs, for: mlsGroupID)
+
+                    context.perform {
+                        completion(.success(()))
+                    }
+
+                } catch {
+                    Logging.eventProcessing.error("Failed to remove member to mls group: \(String(describing: error))")
```
Comment (johnxnguyen) [inline Source/Model/Conversation/ZMConversation+Participants.swift:235]: Better to `warn` than `error` to avoid test issues.

File: Tests/MLS/MLSControllerTests.swift [hunk 4]
```diff
@@ -359,4 +359,98 @@ class MLSControllerTests: ZMConversationTestsBase {
...
+
+        // Mock sending message.
+        mockActionsProvider.sendMessageMocks.append({ message in
+            XCTAssertEqual(message, Data([0, 0, 0, 0]))
+
+            // Mock update event for member join
+            let payload: NSDictionary = [
+                "type": "team.member-join",
+                "data": message
+            ]
```
Comment (johnxnguyen) [inline Tests/MLS/MLSControllerTests.swift:383]: `team.member-join` isn't right. It should be `conversation.member-leave`.

File: Tests/MLS/MLSControllerTests.swift [hunk 5]
```diff
@@ -359,4 +359,98 @@ class MLSControllerTests: ZMConversationTestsBase {
...
+
+            let updateEvent = ZMUpdateEvent(fromEventStreamPayload: payload, uuid: nil)!
+
+            self.mockConversationEventProcessor.processConversationEvents([updateEvent])
+
+            let processConversationEventsCalls = self.mockConversationEventProcessor.calls.processConversationEvents
+            XCTAssertEqual(processConversationEventsCalls.count, 1)
```
Comment (johnxnguyen) [inline Tests/MLS/MLSControllerTests.swift:390]: You shouldn't be calling the mock method directly in your test, it's something that will be called in the `MLSController` when we send the message. You do, however, want to assert that it was called afterwards in `THEN`.

File: Tests/MLS/MLSControllerTests.swift [hunk 6]
```diff
@@ -359,4 +359,98 @@ class MLSControllerTests: ZMConversationTestsBase {
...
+            XCTFail("Unexpected error: \(String(describing: error))")
+        }
+
+        let removeMembersFromConversationCalls = mockCoreCrypto.calls.removeClientsFromConversation
+        XCTAssertEqual(removeMembersFromConversationCalls.count, 1)
+        XCTAssertEqual(removeMembersFromConversationCalls[0].0, mlsGroupID.bytes)
+        XCTAssertEqual(removeMembersFromConversationCalls[0].1, [mlsClientID])
+    }
+
+    func test_AddingMembersToConversation_ThrowsNoClientsToRemove() async {
```
Comment (johnxnguyen) [inline Tests/MLS/MLSControllerTests.swift:409]: The test name says "adding", but the test is for "removing.

File: Tests/MLS/MLSControllerTests.swift [hunk 7]
```diff
@@ -359,4 +359,98 @@ class MLSControllerTests: ZMConversationTestsBase {
...
+            case MLSController.MLSRemoveParticipantsError.noClientsToRemove:
+                break
+
+            default:
+                XCTFail("Unexpected error: \(String(describing: error))")
+            }
+        }
+    }
+
+    func test_AddingMembersToConversation_FailsToSendHandShakeMessage() async {
```
Comment (johnxnguyen) [inline Tests/MLS/MLSControllerTests.swift:429]: The test name says "adding", but the test is for "removing.

File: MLS/MLSController.swift (hunk 2 above)
Comment (ghuMustafa) [reply MLS/MLSController.swift]: Totally agree with it, as here we can avoid our high level interface dependancy to low level, Thanks.

File: Source/Model/Conversation/ZMConversation+Participants.swift (hunk 1 above)
Comment (ghuMustafa) [reply Source/Model/Conversation/ZMConversation+Participants.swift]: Yap I have changed it.

File: Source/Model/Conversation/ZMConversation+Participants.swift (hunk 3 above)
Comment (ghuMustafa) [reply Source/Model/Conversation/ZMConversation+Participants.swift]: Changed.

File: Tests/MLS/MLSControllerTests.swift (hunk 4 above)
Comment (ghuMustafa) [reply Tests/MLS/MLSControllerTests.swift]: Yes it should be conversation.member-leave it occurred while copy past from add participants method, Thanks for identifying.

File: Tests/MLS/MLSControllerTests.swift (hunk 5 above)
Comment (ghuMustafa) [reply Tests/MLS/MLSControllerTests.swift]: Yes, actually it should not be called, I have removed the call and make an assert after the removeMembersFromConversation is called.

File: Tests/MLS/MLSControllerTests.swift (hunk 6 above)
Comment (ghuMustafa) [reply Tests/MLS/MLSControllerTests.swift]: updated.

File: Tests/MLS/MLSControllerTests.swift (hunk 7 above)
Comment (ghuMustafa) [reply Tests/MLS/MLSControllerTests.swift]: updated.

File: MLS/MLSClientID.swift [hunk 8]
```diff
@@ -61,4 +61,9 @@ struct MLSClientID: Equatable {
         self.string = "\(self.userID):\(self.clientID)@\(self.domain)"
     }
 
+    
+    public var bytes: Bytes? {
+        string.base64EncodedString?.base64EncodedBytes
+    }
```
Comment (johnxnguyen) [inline MLS/MLSClientID.swift:67]: Is this necessary? Maybe we can just convert to bytes where we need it?

File: MLS/MLSController.swift [hunk 9]
```diff
@@ -228,6 +229,49 @@ public final class MLSController: MLSControllerProtocol {
...
+        for groupID: MLSGroupID
+    ) async throws {
+
+        guard !clientIds.isEmpty else {
+            throw MLSRemoveParticipantsError.noClientsToRemove
+        }
+        
+        let clientIds =  clientIds.compactMap { $0.bytes }
+
+        let messageToSend = try removeMembers(id: groupID, clientIds: clientIds)
```
Comment (johnxnguyen) [inline MLS/MLSController.swift:252]: What exactly do we need to pass in as client ids?  Raw bytes of the "user:client@domain" id? Or base64 encoded bytes?

File: Source/Utilis/String+Bytes.swift [hunk 10]
```diff
@@ -24,4 +24,8 @@ extension String {
         return Bytes(base64Encoded: self)
     }
 
+    public var base64EncodedString: String? {
+        data(using: .utf8)?.base64EncodedString()
+    }
```
Comment (johnxnguyen) [inline Source/Utilis/String+Bytes.swift:29]: Why is this needed?

File: MLS/MLSController.swift (hunk 9 above)
Comment (ghuMustafa) [reply MLS/MLSController.swift:252]: The Core Crypto expects an array of Raw bytes and ClientId represents `typealias ClientId = [UInt8]` and Here we make raw bytes of `MLSClientID` from String, and to make raw bytes I'm using this method `data(using: .utf8)?.base64EncodedString()`. is it the correct understanding?

File: MLS/MLSClientID.swift (hunk 8 above)
Comment (ghuMustafa) [reply MLS/MLSClientID.swift:67]: when I tried to create bytes with following init method with plan string it return nil as it expects an base64Encoded string.

>    init?(base64Encoded: String) {
        guard let bytes = Data(base64Encoded: base64Encoded)?.bytes else {
            return nil
        }
        self = bytes
    }

Comment (ghuMustafa) [reply MLS/MLSClientID.swift:67]: Yes I got it as we just needed to pass raw data so we no need to convert to base64string.

File: Source/Utilis/String+Bytes.swift (hunk 10 above)
Comment (ghuMustafa) [reply Source/Utilis/String+Bytes.swift]: I just removed the it as we need to pass the Raw bytes instead of base64 encoded bytes.

**Review by johnxnguyen — APPROVED** [review]
